### PR TITLE
add remove links for manifest.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @hellosnow @ansyral @vicancy @superyyrrzz @qinezh

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 1]
```diff
@@ -59,5 +59,41 @@ public static bool AddFile(this Manifest manifest, string sourceFilePath, string
...
             return false;
         }
+
+        public static void RemoveLinks(this Manifest manifest, string manifestFolder)
+        {
+            if (manifest == null)
+            {
+                throw new ArgumentNullException(nameof(manifest));
+            }
+            if (manifestFolder == null)
```
Comment (qinezh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: also check empty string?

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 2]
```diff
@@ -59,5 +59,41 @@ public static bool AddFile(this Manifest manifest, string sourceFilePath, string
             }
             return false;
         }
+
+        public static void RemoveLinks(this Manifest manifest, string manifestFolder)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >RemoveLinks [](start = 27, length = 11)

add some tests?

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 3]
```diff
@@ -59,5 +59,41 @@ public static bool AddFile(this Manifest manifest, string sourceFilePath, string
...
+                {
+                    return;
+                }
+                var fal = FileAbstractLayerBuilder.Default
+                    .ReadFromManifest(manifest, manifestFolder)
+                    .WriteToRealFileSystem(manifestFolder)
+                    .Create();
+                foreach (var rp in ofiList)
+                {
+                    fal.Copy(rp.RelativePath, rp.RelativePath);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >rp.RelativePath, rp.RelativePath [](start = 29, length = 32)

why copy when source == dest?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: reader is manifest with link implements, writer is real file system.

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs (hunk 2 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >RemoveLinks [](start = 27, length = 11)

CentralizeOutputFiles?

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/FileWriterBase.cs [hunk 4]
```diff
@@ -33,6 +33,18 @@ public FileWriterBase(string outputFolder)
 
         #region Help Methods
 
+        protected internal static void EnsureFolder(string folder)
+        {
+            if (string.IsNullOrEmpty(folder))
+            {
+                return;
+            }
+            if (!Directory.Exists(folder))
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/FileWriterBase.cs]: >!Directory.Exists(folder) [](start = 16, length = 25)

why we need this check?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/FileWriterBase.cs]: prevent some one give folder like "c:\"

---
In reply to: [95949599](https://github.com/dotnet/docfx/pull/1172#discussion_r95949599) [](ancestors = 95949599)

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs (hunk 2 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: RealizeLinkedFiles?

---
In reply to: [95947362](https://github.com/dotnet/docfx/pull/1172#discussion_r95947362) [](ancestors = 95947362)

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: Unlink?

---
In reply to: [95950368](https://github.com/dotnet/docfx/pull/1172#discussion_r95950368) [](ancestors = 95950368,95947362)

Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: PhysicalizeLinkFiles?

Comment (chenkennt) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: For naming, is this similar to the "dereference" concept for symbolic links?

**Comment by ansyral** [comment]
:shipit:

**Pull request merged**